examples/autoignition-mpi.py

```diff
@@ -500,9 +502,25 @@ def my_get_timestep(t, dt, state):
                                 op="max")
         return ts_field, cfl, min(t_remaining, dt)
 
+    def limiter(cv, temp=None):
+        spec_lim = make_obj_array([
+            bound_preserving_limiter(discr, cv.species_mass_fractions[i], mmax=1.0)
+            for i in range(nspecies)
+        ])
+
+        kin_energy = 0.5*np.dot(cv.velocity, cv.velocity)
+
+        energy_lim = cv.mass*(
+            gas_model.eos.get_internal_energy(temp, species_mass_fractions=spec_lim)
+            + kin_energy
+        )
+
+        return make_conserved(dim=dim, mass=cv.mass, energy=energy_lim,
+                       momentum=cv.momentum, species_mass=cv.mass*spec_lim)
+
     def my_pre_step(step, t, dt, state):
         cv, tseed = state
-        fluid_state = construct_fluid_state(cv, tseed)
+        fluid_state = construct_fluid_state(limiter(cv, temp=tseed), tseed)
```

Comment:
(I know this is merged already, but I just noticed it.) Should `limiter` be getting compiled somewhere?

Reply:
Good idea. Let me work on that.

Reply:
Note that limiter as called here in this example is in the `pre-step` and does not limit the actual state used by the simulation. Typically I think the limiter is called inside the RHS (`my_rhs`, or whatever), and will be compiled when the RHS is compiled.

Reply:
IIUC, the limiter should operate on the state estimate, not the RHS.

Reply:
Correct.   What I mean is that the limiter should be in the user's RHS function, operating on the state that is passed into the RHS.

Reply:
I'm not sure I agree: I think the limiter should apply to the post-increment state in the time stepper, not the input of the RHS.

@lukeolson, do you agree?

Reply:
@inducer , I agree that it is better to use the limiter to the post-increment of the time stepper (i.e., every sub-iteration of the RK). I am still seen negative values (although not as significant as before) and makes sense that applying only the RHS input (in the pre or post-step) may not be sufficient.

I will update the example this week. Thanks for pointing that out!

Reply:
> @inducer , I agree that it is better to use the limiter to the post-increment of the time stepper (i.e., every sub-iteration of the RK). 

Doing the limiting in the pre/post step callbacks precludes limiting every substep of multi-step timesteppers.  The callbacks are only called once per timestep.

I am not sure that it would matter whether one limits pre, or post-increment.   If the limiting is not performed on the state input to the RHS operator(s), then you have the possibility of feeding a state into the RHS that already has negative values.